infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/LRRollingUpgradeHandler.java

```diff
@@ -47,28 +58,37 @@ public LRRollingUpgradeHandler(ILogReplicationVersionAdapter versionAdapter) {
     }
 
     public boolean isLRUpgradeInProgress(TxnContext txnContext) {
-        if (isClusterAllAtV2) {
+        try {
+            // If LOG_REPLICATION_PLUGIN_VERSION_TABLE exists, it indicates an upgrade from a
+            // previous version was performed. It has since been removed in the current version.
+            txnContext.getTable(LOG_REPLICATION_PLUGIN_VERSION_TABLE);
+
+            if (isClusterAllAtV2) {
+                return false;
+            }
+            String nodeVersion = versionAdapter.getNodeVersion();
+            /**
+             * The ideal way to check the versions is to encapsulate the code version
+             * into Corfu's Layout information so that even when nodes are down
+             * or unresponsive it would be possible to determine if rolling upgrade
+             * is running. But since that is a bigger change we resort to the
+             * boolean check while ensuring migrateData() is idempotent and is a NO-OP
+             * when invoked on a fully upgraded cluster.
+             */
+            String pinnedClusterVersion = versionAdapter.getPinnedClusterVersion(txnContext);
+            boolean isClusterUpgradeInProgress = !nodeVersion.equals(pinnedClusterVersion);
+            if (isClusterUpgradeInProgress) {
+                return true;
+            } // else implies cluster upgrade has completed
+
+            log.info("LRRollingUpgrade upgrade completed to version {}", nodeVersion);
+            migrateData(txnContext);
+            isClusterAllAtV2 = true;
+            return false;
+        } catch (NoSuchElementException e) {
+            log.info("Version table is not present, setup is a new installation");
```

Comment:
Please print the exception as well. From the info message, it is not clear that the exception has been hit.